pgdrive/scene_manager/spawn_manager.py

```diff
@@ -116,33 +124,44 @@ def get_target_vehicle_configs(self, seed=None):
             ret["agent0"] = self.target_vehicle_configs[0]["config"]
         return copy.deepcopy(ret)
 
-    def update(self, vehicles: dict, map):
-        if self.need_update_spawn_places:
-            assert self.initialized
-            self.need_update_spawn_places = False
-            for bid, bp in self.safe_spawn_places.items():
-                lane = map.road_network.get_lane(bp["config"]["spawn_lane_index"])
-                self.safe_spawn_places[bid]["position"] = lane.position(
-                    longitudinal=bp["config"]["spawn_longitude"], lateral=bp["config"]["spawn_lateral"]
-                )
-                for vid in vehicles.keys():
-                    self.confirm_respawn(bid, vid)  # Just assume everyone is all in the same spawn place at t=0.
-
-        for bid, vid_set in self.mapping.items():
-            removes = []
-            for vid in vid_set:
-                if (vid not in vehicles) or (distance_greater(self.safe_spawn_places[bid]["position"],
-                                                              vehicles[vid].position, length=10)):
-                    removes.append(vid)
-            for vid in removes:
-                self.mapping[bid].remove(vid)
-
-    def confirm_respawn(self, spawn_place_id, vehicle_id):
-        self.mapping[spawn_place_id].add(vehicle_id)
-
-    def get_available_spawn_places(self):
+    def step(self):
+        self.spawn_places_used = []
+
+    def get_available_spawn_places(self, pg_world: PGWorld, map):
         ret = {}
-        for bid in self.safe_spawn_places.keys():
-            if not self.mapping[bid]:  # empty
-                ret[bid] = self.safe_spawn_places[bid]
+        for bid, bp in self.safe_spawn_places.items():
+            if bid in self.spawn_places_used:
+                continue
+            lane = map.road_network.get_lane(bp["config"]["spawn_lane_index"])
+            long = bp["config"]["spawn_longitude"]
+            lat = bp["config"]["spawn_lateral"]
+            spawn_point_position = lane.position(longitudinal=long, lateral=lat)
```

Comment:
We don't need to initialize everything here!! We can just put them to the initilization stage! So that we can save time! Since this function is called very frequently, so those variables related to the safe spawn places should not be initliazed  here!

Reply:
Ok, the spawn point will be saved in the safe_spawn_places dict